docs/designs/serving_design.md

```diff
@@ -0,0 +1,101 @@
+
+```python
+def replace_keras_embedding_with_edl_embedding(model)
+    # replace the embedding layer for Sequential and functional API models
+    if isinstance(model, tf.keras.Sequential) or model._is_graph_network:
+        def _clone_function(layer):
+            if type(layer) == keras.layers.Embedding:
+                edl_layer = elastic.layers.Embedding(layer.output_dim)
+                return edl_layer
+            return layer
+        return keras.models.clone_model(model, clone_function=_clone_function)
+    else:
+    # replace embedding attribute for subclass model
+        for attr_name, attr_value in model.__dict__.items():
+            if type(attr_value) == keras.layers.Embedding:
+                setattr(model, attr_name, edl_Embedding(attr_value.output_dim))
+        return model
+```
+
+However, tf.saved_model.save can not export the replaced model to SavedModel. Because ElasticDL.Embedding use tf.py_function to invoke Rpc to interact with the parameter server. It is not mapped to any native TensorFlow op. As a result we choose to save the origin model with native keras embedding layer, replace the embedding parameters with the trained parameters of elasticdl.layers.embedding.
+
+```python
+def restore_keras_embedding_for_subclass(model):
+    for attr_name, attr_value in model.__dict__.items():
+        if type(attr_value) == elasticdl.layers.Embedding:
+            setattr(model, attr_name, keras.layers.Embedding(attr_value.output_dim))
+    return model
+
+def restore_keras_embedding_from_edl_embedding(model):
+    for layer in model.layers:
+        if type(layer) == tf.keras.layers.Embedding:
+            embedding_params = EmbeddingService.get_all_embedding_params(layer)
+            layer.embeddings.assign(embedding_params)
+```
+
+## Execute task to save model with fault-tolerance
+
+We designed the master-worker architecture and task dispatch&recover mechanism in ElasticDL to make the job execution fault tolerant. Please check [overall design](./overall_design.md).\
+For model saving work, we use the same mechanism. After completing all the training/evaluation/prediction tasks, master will generate a SaveModel task and insert it into the todo task queue. The first worker pulling this task will execute the model saving process. Please check the diagram below:
```

Comment:
May need to add more details on what will happen when one of the tasks fail, how do we shard the tasks, etc.

Reply:
I think we don't need shard the SavedModel task. Because we only generate one task of this type. The first worker pulling the task will complete this work.

Reply:
More details are added :)